torch/autograd/gradcheck.py

```diff
@@ -141,11 +141,79 @@ def get_numerical_jacobian(fn, inputs, outputs=None, target=None, eps=1e-3,
     for i, (inp, inp_idx) in enumerate(zip(iter_tensors(target, True), inp_indices)):
-        jacobians += [get_numerical_jacobian_helper(fn, inp, inp_idx, inputs, outputs, eps, grad_out)]
+        jacobians += [get_numerical_jacobian_wrt_specific_input(fn, inp, inp_idx, inputs, outputs, eps, grad_out)]
     return jacobians
 
 
-def prepped_input(input, maybe_perturbed_input):
+def compute_numerical_gradient(fn, entry, v, nbhd_checks_fn):
+    # Performs finite differencing by perturbing `entry` in-place by `v` and
+    # returns the gradient of each of the outputs wrt to x at idx.
+    norm_v = -1j * v if isinstance(v, complex) else v
+
+    orig = entry.clone()
+    entry.copy_(orig - v)
+    outa = fn()
+    entry.copy_(orig + v)
+    outb = fn()
+    entry.copy_(orig)
+
+    def compute(a, b):
+        nbhd_checks_fn(a, b)
+        ret = (b - a) / (2 * norm_v)
+        return ret.detach().reshape(-1)
+
+    return tuple(compute(a, b) for (a, b) in zip(outa, outb))
+
+
+def compute_numerical_jacobian_cols(jvp_fn, delta, input_is_complex, grad_out) -> List[torch.Tensor]:
+    # Computing the jacobian only works for pure real or pure imaginary delta
+    # for details on the algorithm used here, refer:
+    # Section 3.5.3 https://arxiv.org/pdf/1701.00392.pdf
+    # s = fn(z) where z = x for real valued input
+    # and z = x + yj for complex valued input
+    jacobians_cols: List[torch.Tensor] = []
+    ds_dx_tup = jvp_fn(delta)
+
+    if input_is_complex:            # C -> C, C -> R
+        ds_dy_tup = jvp_fn(delta * 1j)
+        for ds_dx, ds_dy in zip(ds_dx_tup, ds_dy_tup):
+            if not ds_dx.is_complex() and isinstance(grad_out, complex):
+                # placeholder if grad_out is complex but output is not
+                jacobians_cols.append(torch.zeros_like(ds_dx))
+                continue
+            # conjugate wirtinger derivative
+            conj_w_d = 0.5 * (ds_dx + ds_dy * 1j)
+            # wirtinger derivative
+            w_d = 0.5 * (ds_dx - ds_dy * 1j)
+            jacobians_cols.append(grad_out.conjugate() * conj_w_d + grad_out * w_d.conj())
+    else:
+        for ds_dx in ds_dx_tup:
+            if ds_dx.is_complex():  # R -> C
+                # w_d = conj_w_d = 0.5 * ds_dx
+                # dL_dz_conj = 0.5 * [grad_out.conj() * ds_dx + grad_out * ds_dx.conj()]
+                #            = 0.5 * [grad_out.conj() * ds_dx + (grad_out.conj() * ds_dx).conj()]
+                #            = 0.5 * 2 * real(grad_out.conj() * ds_dx)
+                #            = real(grad_out.conj() * ds_dx)
+                jacobians_cols.append(torch.real(grad_out.conjugate() * ds_dx))
+            else:                   # R -> R
+                if isinstance(grad_out, complex):
+                    # placeholder if grad_out is complex but output is not
+                    jacobians_cols.append(torch.zeros_like(ds_dx))
+                    continue
```

Comment:
Previously we appended None here, but then I realized that line 212 wouldn't work properly (assigning NoneType to the column of a Tensor). The tests were still passing though because the code path was never reached, i.e., we didn't have any functions that take real input and have both complex and real output.